src/java/org/apache/cassandra/db/counters/CachedCounterLockManager.java

```diff
@@ -0,0 +1,154 @@
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.StampedLock;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+
+/**
+ * Implemetation of {@link CounterLockManager} that uses a cache of locks.
+ * Note: this implemetation tries to reduce the chance of having two counters lock each other, but as the counters
+ * are identified by the hash of the primary key of the row, it is still possible to
+ * have some cross-counter contention for counters with different primary keys but the same hash.
+ *
+ * This code is copied from
+ * <a href="https://github.com/diennea/herddb/blob/master/herddb-utils/src/main/java/herddb/utils/LocalLockManager.java">LocalLockManager</a> from the HerdDB project (Apache 2 licensed).
+ *
+ */
+public class CachedCounterLockManager implements CounterLockManager
+{
+    private final static int EXPECTED_CONCURRENCY = DatabaseDescriptor.getConcurrentCounterWriters() * 16;
+    private final ConcurrentMap<Integer, RefCountedStampedLock> locks = new ConcurrentHashMap<>(EXPECTED_CONCURRENCY);
+
+    @Override
+    public List<LockHandle> grabLocks(Iterable<Integer> keys)
+    {
+        List<LockHandle> result = new ArrayList<>();
+        for (Integer key : keys)
+            result.add(makeLockForKey(key));
+        return result;
+    }
+
+    private StampedLock makeLock()
+    {
+        return new StampedLock();
+    }
+
+    private LockHandleImpl makeLockForKey(Integer key)
+    {
+        RefCountedStampedLock instance = locks.compute(key, (k, existing) -> {
+            if (existing != null)
+            {
+                existing.count++;
+                return existing;
+            }
+            else
+            {
+                return new RefCountedStampedLock(makeLock(), 1);
+            }
+        });
+        return new LockHandleImpl(key, instance);
+    }
+
+    private void returnLockForKey(RefCountedStampedLock instance, Integer key) throws IllegalStateException
+    {
+        locks.compute(key, (Integer t, RefCountedStampedLock u) -> {
+            if (instance != u)
+            {
+                throw new IllegalStateException("trying to release un-owned lock");
+            }
+            if (--u.count == 0)
+            {
+                return null;
+            }
+            else
+            {
+                return u;
+            }
+        });
+    }
+
+    @Override
+    public boolean hasNumKeys()
+    {
+        return true;
+    }
+
+    @Override
+    public int getNumKeys()
+    {
+        return locks.size();
+    }
+
+    /**
+     * This class is not thread safe, it is expected to be used by a single thread.
+     */
+    private class LockHandleImpl implements LockHandle
+    {
+        public long stamp;
+        public final Integer key;
+        public final RefCountedStampedLock handle;
+
+        public LockHandleImpl(Integer key, RefCountedStampedLock handle)
```

Comment:
Please make this private.
public constructor suggests that the handle may be shared, whereas the `LockHandleImpl` really owns it.
This aspect would be best expressed if the `LockHandleImpl` was in full control of the lifetimes (i.e. created the handle itself), but I guess it's fine as-is in a private class

Reply:
make sense, fixed